src/libraries/System.Net.NameResolution/tests/FunctionalTests/TelemetryTest.cs

```diff
@@ -123,17 +110,118 @@ public static void EventSource_GetHostEntryForIP_LogsStartStop()
+            }
+            else
+            {
+                Assert.DoesNotContain(events, e => e.Event.EventName == "ResolutionFailed");
+            }
+        }
 
-                EventWrittenEventArgs[] stops = events.Where(e => e.EventName == "ResolutionStop").ToArray();
-                Assert.Equal(12, stops.Length);
+        [ConditionalFact(typeof(RemoteExecutor), nameof(RemoteExecutor.IsSupported))]
+        public static void ResolutionsWaitingOnQueue_ResolutionStartCalledBeforeEnqueued()
+        {
+            // Some platforms (non-Windows) don't have proper support for GetAddrInfoAsync.
+            // Instead we perform async-over-sync with a per-host queue.
+            // This test ensures that ResolutionStart events are written before waiting on the queue.
+
+            // We do this by blocking the first ResolutionStart event.
+            // If the event was logged after waiting on the queue, the second request would never complete.
+            RemoteExecutor.Invoke(async () =>
+            {
+                using var listener = new TestEventListener("System.Net.NameResolution", EventLevel.Informational);
+                listener.AddActivityTracking();
+
+                TaskCompletionSource firstResolutionStart = new();
+                TaskCompletionSource secondResolutionStop = new();
+
+                List<(string EventName, Guid ActivityId)> events = new();
+
+                await listener.RunWithCallbackAsync(e =>
+                {
+                    if (e.EventName == "ResolutionStart" || e.EventName == "ResolutionStop")
+                    {
+                        Assert.NotEqual(Guid.Empty, e.ActivityId);
+                        events.Add((e.EventName, e.ActivityId));
+                    }
+
+                    if (e.EventName == "ResolutionStart" && firstResolutionStart.TrySetResult())
+                    {
+                        secondResolutionStop.Task.Wait();
+                    }
+                },
+                async () =>
+                {
+                    Task first = DoResolutionAsync();
+
+                    await firstResolutionStart.Task.WaitAsync(TimeSpan.FromSeconds(30));
+
+                    Task second = DoResolutionAsync();
+
+                    await Task.WhenAny(first, second).WaitAsync(TimeSpan.FromSeconds(30));
+                    Assert.False(first.IsCompleted);
+
+                    await second;
+                    secondResolutionStop.SetResult();
+
+                    await first.WaitAsync(TimeSpan.FromSeconds(30));
+
+                    static Task DoResolutionAsync()
+                    {
+                        return Task.Run(async () =>
+                        {
+                            try
+                            {
+                                await Dns.GetHostAddressesAsync("microsoft.com");
+                            }
+                            catch (AggregateException ex) when (ex.InnerException is not XunitException)
```

Comment:
Why would we get an AggregateException here?

Reply:
`XunitException` could be thrown if an Assert in the event callback is hit. Specifically this one:
https://github.com/dotnet/runtime/blob/eec1b850d086f117c56d18776981970eeb7921fd/src/libraries/System.Net.NameResolution/tests/FunctionalTests/TelemetryTest.cs#L176

The exception should be just `Exception` and not `AggregateException` here though, I'll change it

Reply:
You mean exceptions from EventListeners are propagating out through the WriteEvent code?!

Reply:
Yes, they are wrapped in `EventSourceException` and rethrown, unless the `EventSource` was created with the `EventSourceSettings.ThrowOnEventWriteErrors` flag removed.

Reply:
I'm very surprised by that.  I'm concerned we haven't taken that into account in pretty much any of our event sources, e.g. that suggests `ArrayPool<byte>.Shared.Return(array)` could throw arbitrary exceptions from any listener in the process.

@noahfalk, @brianrob, this is expected?

@GrabYourPitchforks, @bartonjs, I've pushed for a certain pattern of ArrayPool consumption, where when Return is used in a try and then also in a finally, the inner code first stores the array into a temporary, nulls out the field, and then returns the temporary, but that pattern isn't followed consistently.  Assuming this EventSource behavior isn't going to change, I think we need to audit all usage, as code like https://github.com/dotnet/runtime/blob/5399ee59e55a75a850011908a0f96d7abb8a91c7/src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs#L745-L765 is now capable of double-returns to the pool in the face of an arbitrary EventListener throwing an exception.

Reply:
I had assumed that `ThrowOnEventWriteErrors` was set by default, but in reality it just doesn't apply to exceptions thrown by `EventListener`s (those are always rethrown). That feels like a bug to me.

Reply:
This does not sound right to me.  ETW, EventPipe and EventListener writes are wrapped in a try/catch to ensure that they don't cause exceptions to flow back through an event write call site.  By default, `ThrowOnEventWriteErrors == false`, which means that you'll get an event written with the failure, and if that fails to write, EventSource will just give up.  If you turn on `ThrowOnEventWriteErrors`, then you'll get exceptions thrown through event write call sites, but this is something that the EventSource itself has to opt-into.

Reply:
That sounds right to me for what we would want to happen, but rn it's not what the implementation does.

The `ThrowOnEventWriteErrors` check is skipped here: https://github.com/dotnet/runtime/blob/4e47e8cfa22eb9c5be502d96aa7e5d7418dcff84/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs#L2000-L2001

Likely just missing the fact that `DispatchToAllListeners` already catches and throws the same exception here: https://github.com/dotnet/runtime/blob/4e47e8cfa22eb9c5be502d96aa7e5d7418dcff84/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs#L2132

Reply:
Simple repro:
```C#
using System.Diagnostics.Tracing;

using var listener = new MyListener();

try
{
    MySource.Log.SomethingHappened();
}
catch (Exception e)
{
    Console.WriteLine($"Caught exception: {e}");
}

class MySource : EventSource
{
    public static readonly MySource Log = new();

    [Event(1)]
    public void SomethingHappened() => WriteEvent(1);
}

class MyListener : EventListener
{
    public MyListener() => EnableEvents(MySource.Log, EventLevel.LogAlways);

    protected override void OnEventWritten(EventWrittenEventArgs eventData) =>
        throw new Exception("Uh oh... you shouldn't see this");
}
```
prints:
```
Caught exception: System.Diagnostics.Tracing.EventSourceException: An error occurred when writing to a listener.
 ---> System.Exception: Uh oh... you shouldn't see this
   at MyListener.OnEventWritten(EventWrittenEventArgs eventData) in C:\Users\stoub\source\repos\ConsoleApp28\ConsoleApp28\Program.cs:line 27
   at System.Diagnostics.Tracing.EventSource.DispatchToAllListeners(EventWrittenEventArgs eventCallbackArgs) in System.Private.CoreLib.dll:token 0x60066f3+0x23
   --- End of inner exception stack trace ---
   at System.Diagnostics.Tracing.EventSource.DispatchToAllListeners(EventWrittenEventArgs eventCallbackArgs) in System.Private.CoreLib.dll:token 0x60066f3+0x5f
   at System.Diagnostics.Tracing.EventSource.WriteToAllListeners(EventWrittenEventArgs eventCallbackArgs, Int32 eventDataCount, EventData* data) in System.Private.CoreLib.dll:token 0x60066f2+0x118
   at System.Diagnostics.Tracing.EventSource.WriteEventWithRelatedActivityIdCore(Int32 eventId, Guid* relatedActivityId, Int32 eventDataCount, EventData* data) in System.Private.CoreLib.dll:token 0x60066e0+0x1d9
   at System.Diagnostics.Tracing.EventSource.WriteEvent(Int32 eventId) in System.Private.CoreLib.dll:token 0x60066ce+0x0
   at MySource.SomethingHappened() in C:\Users\stoub\source\repos\ConsoleApp28\ConsoleApp28\Program.cs:line 19
   at <Program>$.<Main>$(String[] args) in C:\Users\stoub\source\repos\ConsoleApp28\ConsoleApp28\Program.cs:line 7
```

Reply:
Repros on .NET Framework 4.8 as well:
```C#
using System;
using System.Diagnostics.Tracing;

class Program
{
    static void Main()
    {
        using (var listener = new MyListener())
        {
            try
            {
                MySource.Log.SomethingHappened();
            }
            catch (Exception e)
            {
                Console.WriteLine($"Caught exception: {e}");
            }
        }
    }
}

class MySource : EventSource
{
    public static readonly MySource Log = new MySource();

    [Event(1)]
    public void SomethingHappened() => WriteEvent(1);
}

class MyListener : EventListener
{
    public MyListener() => EnableEvents(MySource.Log, EventLevel.LogAlways);

    protected override void OnEventWritten(EventWrittenEventArgs eventData) =>
        throw new Exception("Uh oh... you shouldn't see this");
}
```
printing:
```
Caught exception: System.Diagnostics.Tracing.EventSourceException: An error occurred when writing to a listener. ---> System.Exception: Uh oh... you shouldn't see this
   at MyListener.OnEventWritten(EventWrittenEventArgs eventData) in C:\Users\stoub\source\repos\ConsoleApp29\ConsoleApp29\Program.cs:line 35
   at System.Diagnostics.Tracing.EventSource.DispatchToAllListeners(Int32 eventId, Guid* childActivityID, EventWrittenEventArgs eventCallbackArgs)
   --- End of inner exception stack trace ---
   at System.Diagnostics.Tracing.EventSource.DispatchToAllListeners(Int32 eventId, Guid* childActivityID, EventWrittenEventArgs eventCallbackArgs)
   at System.Diagnostics.Tracing.EventSource.WriteToAllListeners(Int32 eventId, Guid* activityID, Guid* childActivityID, Object[] args)
   at System.Diagnostics.Tracing.EventSource.WriteToAllListeners(Int32 eventId, Guid* activityID, Guid* childActivityID, Int32 eventDataCount, EventData* data)
   at System.Diagnostics.Tracing.EventSource.WriteEventWithRelatedActivityIdCore(Int32 eventId, Guid* relatedActivityId, Int32 eventDataCount, EventData* data)
   at System.Diagnostics.Tracing.EventSource.WriteEvent(Int32 eventId)
   at MySource.SomethingHappened() in C:\Users\stoub\source\repos\ConsoleApp29\ConsoleApp29\Program.cs:line 27
   at Program.Main() in C:\Users\stoub\source\repos\ConsoleApp29\ConsoleApp29\Program.cs:line 12
```

Reply:
You're right - it comes from here: https://github.com/dotnet/runtime/blob/4e47e8cfa22eb9c5be502d96aa7e5d7418dcff84/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs#L2130.  That is not what I would have expected.  I am wondering if there was a historical desire to not hide errors thrown in-process, but I can't think of a good reason why we would want to do this.  This is arguably a compat break, but I feel like we should be changing this behavior to not throw.  I don't think EventListeners should be throwing, but I think it's worse to propagate those exceptions up to to the EventSource unless the EventSource has asked for this.  @noahfalk what do you think?

Reply:
https://github.com/dotnet/runtime/pull/56229

Reply:
Opened #56232 for the EventSource issue.

I removed the reliance on this behavior from this PR.

Reply:
I'm with Brian on this. This is not the behavior I would expect from the in-process scenario. I'll review @MihaZupan's PR.

I did a quick audit of all write methods in EventSource.cs and TraceLoggingEventSource.cs and all the write methods seem to have try catches that special case `EventSourceException`:

https://github.com/dotnet/runtime/blob/4e47e8cfa22eb9c5be502d96aa7e5d7418dcff84/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs#L1998-L2004

Where the `ThrowEventSourceException` method checks the config and only rethrows if this option is true.

So the exception from EventListener is bubbling up because the EventListener failure (exception thrown in `OnEventWritten`) is being rethrown as an `EventSourceExcpetion`.

It looks like the intended behavior is if EventSource itself encounters a problem, it will throw, but if it encounters and issue while writing that isn't its fault it rethrows it.

I don't think that matches the intended behavior based on this comment:
https://github.com/dotnet/runtime/blob/4e47e8cfa22eb9c5be502d96aa7e5d7418dcff84/src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs#L684-L697

Reply:
> I don't think EventListeners should be throwing, but I think it's worse to propagate those exceptions up to to the EventSource unless the EventSource has asked for this. @noahfalk what do you think?

I agree with you all, if the EventSource didn't opt in to raising Exceptions then we shouldn't allow the Exceptions to propagate through and we can't rely on EventListeners to be well behaved.